core/scale/big_fixed_integers.hpp

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_BIG_FIXED_INTEGERS_HPP
+#define KAGOME_BIG_FIXED_INTEGERS_HPP
+
+#include <type_traits>
+
+#include <boost/multiprecision/cpp_int.hpp>
+
+#include "common/int_serialization.hpp"
+
+namespace scale {
+
+  /**
+   * Wrapper for an integer to make distinct integer aliases
+   * @tparam Tag
+   * @tparam T
+   */
+  template <typename Tag, typename T>
+  struct IntWrapper {
+    using Self = IntWrapper<Tag, T>;
+
+    T &operator*() {
+      return number;
+    }
+
+    const T &operator*() const {
+      return number;
+    }
+
+    bool operator==(Self const &other) const {
+      return **this == *other;
+    }
+
+    T number;
+  };
+
+  struct FixedTag {};
+
+  // an integer intended to be encoded with fixed length
+  template <typename T>
+  using Fixed = IntWrapper<FixedTag, T>;
+
+  struct CompactTag {};
+
+  // an integer intended to be encoded with compact encoding
+  template <typename T>
+  using Compact = IntWrapper<CompactTag, T>;
+
+  // intended to wrap unwrapped integer arguments to encoding functions to
+  // indicated they should be encoded as fixed-length integers
+  template <typename T>
+  struct EncodeAsFixed {
+    explicit EncodeAsFixed(T &number) : number{number} {}
+    T &number;
+  };
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_decoder_stream>,
+            typename = std::enable_if_t<std::numeric_limits<N>::is_integer>>
+  Stream &operator>>(Stream &stream, Fixed<N> &fixed) {
+    return stream >> EncodeAsFixed{*fixed};
+  }
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_encoder_stream>,
+            typename = std::enable_if_t<std::numeric_limits<N>::is_integer>>
+  Stream &operator<<(Stream &stream, Compact<N> compact) {
+    scale::CompactInteger n = compact;
+    stream << n;
+    return stream;
+  }
+
+  template <typename Stream,
+            typename N,
+            typename = std::enable_if_t<Stream::is_decoder_stream>,
+            typename = std::enable_if_t<std::numeric_limits<N>::is_integer>>
+  Stream &operator>>(Stream &stream, Compact<N> &compact) {
+    scale::CompactInteger n;
+    stream >> n;
+    compact.number = n.convert_to<N>();
+    return stream;
+  }
+
+  // intended to wrap unwrapped integer arguments to encoding functions to
+  // indicated they should be encoded with compact encoding
+  template <typename T>
+  struct EncodeAsCompact {
+    explicit EncodeAsCompact(T &number) : number{number} {}
+    T &number;
+  };
+
+  // marker for big integer types
+  template <typename T>
+  struct BigIntegerTraits {
+    static constexpr bool value = false;
+
+    static constexpr size_t BIT_SIZE = sizeof(T) * 8;
+  };
```

Comment:
Require explicit trait implementations by deleting default.
```suggestion
  struct BigIntegerTraits;
```